inc/search.php

```diff
@@ -0,0 +1,300 @@
+<?php
+
+namespace HM_Handbook\Search;
+
+use \WP_REST_Server;
+use \WP_REST_Request;
+use \WP_Query;
+
+add_action( 'wp_enqueue_scripts', function() {
+
+	wp_localize_script( 'hm-handbook', 'HMHandbookSearchSettings', [
+		'api_endpoint' => rest_url( 'hm-handbook/v1/search' ),
+		'api_nonce'    => wp_create_nonce( 'wp_rest' ),
+	] );
+
+	add_action( 'wp_footer', __NAMESPACE__ . '\\search_templates' );
+
+}, 20 );
+
+add_action( 'rest_api_init', function() {
+
+	register_rest_route( 'hm-handbook/v1', '/search', array(
+		'callback'     => __NAMESPACE__ . '\\search_request_callback',
+		'methods'      => WP_REST_Server::READABLE,
+		'args' => [
+			'query' => [
+				'sanitize_callback' => 'sanitize_text_field',
+			],
+		],
+	) );
+
+} );
+
+class Result {
+	public $title;
+	public $excerpt;
+	public $site;
+	public $author;
+	public $date;
+	public $url;
+	public $type;
+}
+
+function search_request_callback( WP_Rest_Request $request ) {
+
+	$query = $request->get_param( 'query' );
+
+	if ( function_exists( 'hmn_hmes_search_items' ) ) {
+		$data = elastic_search_request( $query );
+	} else {
+		$data = fallback_search_request( $query );
+	}
+
+	return rest_ensure_response( $data );
+
+}
+
+/**
+ * Do the elastic search request.
+ *
+ * @param string $query Query string.
+ *
+ * @return array Array of HM_Handbook\Search\Result objects.
+ */
+function elastic_search_request( $query ) {
+
+	$args  = array();
+	$match = preg_match_all( '~(\S+): ?(\S+)~i', $query, $matches );
+
+	if ( $match && ! empty( $matches[0] ) ) {
+
+		foreach ( $matches[0] as $key_val ) {
+			$query                  = preg_replace( '~' . preg_quote( $key_val ) . '~', '', $query );
+			$exploded               = explode( ':', str_replace( ' ', '', $key_val ) );
+			$args[ $exploded[ 0 ] ] = strpos( $exploded[1], ',' ) ? explode( ',', $exploded[1] ) : $exploded[1];
+		}
+	}
+
+	if ( ! empty( $args['before'] ) ) {
+		// If only a number is supplied, assume year
+		$args['before'] = is_numeric( $args['before'] ) ? strtotime( $args['before'] . '-01-01' ) :   strtotime( $args['before'] );
+	} else {
+		$args['before'] = null;
+	}
+
+	if ( ! empty( $args['after'] ) ) {
+		// If only a number is supplied, assume year
+		$args['after'] = is_numeric( $args['after'] ) ? strtotime( $args['after'] . '-01-01' ) :   strtotime( $args['after'] );
+	}  else {
+		$args['after'] = null;
+	}
+
+	if ( ! empty( $args['user'] ) && ! is_numeric( $args['user'] )  ) {
+
+		global $wpdb;
+
+		$r = $wpdb->get_var( $wpdb->prepare( "SELECT ID FROM $wpdb->users WHERE user_login = %s OR display_name LIKE %s ", $args['user'], '%' . $args['user'] . '%' ) );
```

Comment:
I'm not too familiar either, and I copy/pasted from the Hub site implementation. 

I think there are some fancy search thing supported such as `user:Tom` to fetch posts by a single user. Or `site:` or restricting by date.

Actually, whilst all of these things are nice and work particularly well on the Hub, i'm not sure how useful they are on the handbook. Maybe I just remove it all? If it stays, we need some way to show what filters are available.